### PR TITLE
[Loader][Serializer]: Add Function References Instructions and their tests

### PR DESCRIPTION
**Pull request by elhewaty**
Signed-off-by: Mohamed Atef <mohamedatef1698@gmail.com>

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by elhewaty** [comment]
CC @hydai

File: test/loader/serializeInstructionTest.cpp [hunk 1]
```diff
@@ -666,4 +666,119 @@ TEST(SerializeInstructionTest, SerializeConstInstruction) {
...
+  std::vector<uint8_t> Output;
+  std::vector<WasmEdge::AST::Instruction> Instructions;
+
+  // 12. Test typed function references control flow instructions.
+  //
+  //   1.  Serialize br_on_null instruction.
+  //   2.  Serialize br_on_non_null instruction.
+
+  Conf.addProposal(WasmEdge::Proposal::FunctionReferences);
+  WasmEdge::AST::Instruction Br_on_null(WasmEdge::OpCode::Br_on_null);
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:681]: Please follow the naming pattern above (i.e., `BrOnNull` for example).

File: test/loader/serializeInstructionTest.cpp [hunk 2]
```diff
@@ -666,4 +666,119 @@ TEST(SerializeInstructionTest, SerializeConstInstruction) {
   };
   EXPECT_EQ(Output, Expected);
 }
+
+TEST(SerializeInstructionTest, SerializeFuncRefControlInstruction) {
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:670]: The `br_on_null` and `br_on_non_null` instructions are both control instructions.
Move these into the `SerializeBrControlInstruction` test.

File: test/loader/serializeInstructionTest.cpp [hunk 3]
```diff
@@ -666,4 +666,119 @@ TEST(SerializeInstructionTest, SerializeConstInstruction) {
...
+  std::vector<uint8_t> Expected;
+  std::vector<uint8_t> Output;
+  std::vector<WasmEdge::AST::Instruction> Instructions;
+
+  // 12. Test typed function references control flow instructions.
+  //
+  //   1.  Serialize br_on_null instruction.
+  //   2.  Serialize br_on_non_null instruction.
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:678]: In addition, you can add the fail case (`EXPECT_FALSE`) for testing these instructions without the typed function reference proposal config.

File: test/loader/serializeInstructionTest.cpp [hunk 4]
```diff
@@ -666,4 +666,119 @@ TEST(SerializeInstructionTest, SerializeConstInstruction) {
...
+  Output = {};
+  EXPECT_TRUE(Ser.serializeSection(createCodeSec(Instructions), Output));
+  Expected[5] = 0xD6U; // OpCode Br_on_non_null.
+  EXPECT_EQ(Output, Expected);
+
+  // Remove FunctionReferences Proposal
+  Conf.removeProposal(WasmEdge::Proposal::FunctionReferences);
+}
+
+TEST(SerializeInstructionTest, SerializeFuncRefCallInstruction) {
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:711]: These are call instructions. Move them into the `SerializeCallInstruction` test.

File: test/loader/serializeInstructionTest.cpp [hunk 5]
```diff
@@ -666,4 +666,119 @@ TEST(SerializeInstructionTest, SerializeConstInstruction) {
...
+  EXPECT_TRUE(Ser.serializeSection(createCodeSec(Instructions), Output));
+  Expected[5] = 0x15U; // OpCode Return_call_ref.
+  EXPECT_EQ(Output, Expected);
+
+  // Remove the added Proposals
+  Conf.removeProposal(WasmEdge::Proposal::FunctionReferences);
+  Conf.removeProposal(WasmEdge::Proposal::TailCall);
+}
+
+TEST(SerializeInstructionTest, SerializeRefAsNonNullInstruction) {
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:756]: Move into `SerializeReferenceInstruction` test.

**Review by q82419 — CHANGES_REQUESTED** [review]
Well done. Please follow the requested changes. Thanks!

**Comment by elhewaty** [comment]
@q82419, updated.

File: test/loader/serializeInstructionTest.cpp [hunk 6]
```diff
@@ -215,9 +215,13 @@ TEST(SerializeInstructionTest, SerializeBrControlInstruction) {
   // 3. Test branch control instructions.
   //
   //   1.  Serialize valid label index.
+  //   2.  Serialize without Func-Ref proposal.
 
+  Conf.addProposal(WasmEdge::Proposal::FunctionReferences);
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:220]: This line should be right before the 2. test.

File: test/loader/serializeInstructionTest.cpp [hunk 7]
```diff
@@ -215,9 +215,13 @@ TEST(SerializeInstructionTest, SerializeBrControlInstruction) {
   // 3. Test branch control instructions.
   //
   //   1.  Serialize valid label index.
+  //   2.  Serialize without Func-Ref proposal.
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:218]: 2. Serialize Br_on_null instruction with Func-Ref proposal.
3. Serialize Br_on_non_null instruction with Func-Ref proposal.
4. Serialize invalid Br_on_non_null instruction without Func-Ref proposal.

File: test/loader/serializeInstructionTest.cpp [hunk 8]
```diff
@@ -315,11 +339,18 @@ TEST(SerializeInstructionTest, SerializeCallControlInstruction) {
...
+  //   5.  Serialize Return_call_ref instruction with valid type and table
+  //   6.  Serialize Return_call_ref without Func-Ref proposal.
 
   WasmEdge::AST::Instruction Call(WasmEdge::OpCode::Call);
   WasmEdge::AST::Instruction CallIndirect(WasmEdge::OpCode::Call_indirect);
+  WasmEdge::AST::Instruction CallRef(WasmEdge::OpCode::Call_ref);
+  WasmEdge::AST::Instruction ReturnCallRef(WasmEdge::OpCode::Return_call_ref);
   WasmEdge::AST::Instruction End(WasmEdge::OpCode::End);
 
+  Conf.addProposal(WasmEdge::Proposal::FunctionReferences);
```
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:352]: As above, add right before 4. test.

**Review by q82419 — CHANGES_REQUESTED** [review]

**Review by q82419 — APPROVED** [review]

**Comment by elhewaty** [comment]
@q82419 when will you merge it?

**Pull request merged**